solidity/contracts/vault/TBTCOptimisticMinting.sol

```diff
@@ -0,0 +1,377 @@
+            "Optimistic minting delay has not passed yet"
+        );
+
+        Deposit.DepositRequest memory deposit = bridge.deposits(depositKey);
+        require(deposit.sweptAt == 0, "The deposit is already swept");
+
+        // Bridge, when sweeping, cuts a deposit treasury fee and splits
+        // Bitcoin miner fee for the sweep transaction evenly between the
+        // depositors in the sweep.
+        //
+        // When tokens are optimistically minted, we do not know what the
+        // Bitcoin miner fee for the sweep transaction will look like.
+        // The Bitcoin miner fee is ignored. When sweeping, the miner fee is
+        // subtracted so the optimisticMintingDebt may stay non-zero after the
+        // deposit is swept.
+        //
+        // This imbalance is supposed to be solved by a donation to the Bridge.
+        uint256 amountToMint = deposit.amount - deposit.treasuryFee;
+        _mint(deposit.depositor, amountToMint);
+        optimisticMintingDebt[deposit.depositor] += amountToMint;
+
+        delete pendingOptimisticMints[depositKey];
+
+        emit OptimisticMintingFinalized(
+            msg.sender,
+            deposit.depositor,
+            amountToMint,
+            fundingTxHash,
+            fundingOutputIndex,
+            depositKey
+        );
+    }
+
+    /// @notice Allows a Guardian to cancel optimistic minting request. The
+    ///         following conditions must be met:
+    ///         - The optimistic minting request for the given deposit exists.
+    ///         - The optimistic minting request for the given deposit has not
+    ///           been finalized yet.
+    /// @dev Guardians must validate the following conditions for every deposit
+    ///      for which the optimistic minting was requested:
+    ///      - The deposit happened on Bitcoin side and it has enough
+    ///        confirmations.
+    ///      - The optimistic minting has been requested early enough so that
+    ///        the wallet has enough time to sweep the deposit.
+    ///      - The wallet is an active one and it does perform sweeps or it will
+    ///        perform sweeps once the sweeps are activated.
+    function cancelOptimisticMint(
+        bytes32 fundingTxHash,
+        uint32 fundingOutputIndex
+    ) external onlyGuardian {
+        uint256 depositKey = calculateDepositKey(
+            fundingTxHash,
+            fundingOutputIndex
+        );
+
+        require(
+            pendingOptimisticMints[depositKey] > 0,
+            "Optimistic minting not requested of already finalized"
+        );
+
+        delete pendingOptimisticMints[depositKey];
+
+        emit OptimisticMintingCancelled(
+            msg.sender,
+            fundingTxHash,
+            fundingOutputIndex,
+            depositKey
+        );
+    }
+
+    /// @notice Adds the address to the Minter set.
+    function addMinter(address minter) external onlyOwner {
```

Comment:
`removeMinter` has `onlyOwnerOrGuardian` so that a Guardian can remove an errant minter, see [this](https://github.com/keep-network/tbtc-v2/pull/424#discussion_r1042313896) discussion. We may have a situation when an errant Minter keeps requesting optimistic minting for deposits that did not happen on Bitcoin in a loop. Then, Guardians need to keep canceling these deposit requests in a loop and it becomes a game of who is faster and who has more ether. `removeMinter` will hopefully never be used by Guardians.

In the case of `addMinter`, we do not want any of the Guardians to be able to add Minters. The only role of the Guardian is to watch what Minters are doing and potentially react, not to decide who is a Minter. It is best for security when these two sets are separated.